kernel/smp.c

```diff
@@ -48,3 +50,56 @@ void _smp_global_unlock(unsigned int key)
 	recursive_count = 0;
 	k_spin_unlock(&global_spinlock, sk);
 }
+
+extern k_thread_stack_t _interrupt_stack1[];
+extern k_thread_stack_t _interrupt_stack2[];
+extern k_thread_stack_t _interrupt_stack3[];
+
+#ifdef CONFIG_SMP
+static void _smp_init_top(int key, void *arg)
+{
+	atomic_t *start_flag = arg;
+
+	/* Wait for the signal to begin scheduling */
+	do {
+		k_busy_wait(100);
+	} while (!atomic_get(start_flag));
+
+	/* Switch out of a dummy thread.  Trick cribbed from the main
+	 * thread init.  Should probably unify implementations.
+	 */
+	struct k_thread dummy_thread = {
+		.base.user_options = K_ESSENTIAL,
+		.base.thread_state = _THREAD_DUMMY,
+	};
+
+	_arch_curr_cpu()->current = &dummy_thread;
+	_Swap(irq_lock());
+
+	CODE_UNREACHABLE;
+}
+#endif
+
+void smp_init(void)
+{
+	atomic_t start_flag;
+
+	atomic_clear(&start_flag);
```

Comment:
Just for my understanding, why we need atomic here?

Reply:
We don't on ESP-32 or x86.  In theory on some architecture we might if there are cache coherence rules..  Beyond that it's just form adherence.  Spinning on a raw int from a separate CPU just feels wrong.